### DISCUSSION TITLE
LDM Notes for May 3rd, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7168] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-05-03.md

## Agenda

- Inline Arrays
- Primary constructors
    - Attributes on captured parameters
    - Warning for shadowing base members
- Collection literal natural type

**Comment by TahirAhmadov** [comment]
Regarding collection literal natural type, there are two important points that were already discussed in the main issue:
- Immutable collections are not very useful in the scenario where a natural type is needed at all, specifically - `var` locals. Why else would you assign a collection to a local, other than to modify it? Yes, in some cases the same collection can be passed as an argument to multiple method calls, which would save allocations, but those methods expect a specific parameter type, so it's to be expected `var` is frequently not usable at all for this.
- ~The `foreach` scenario is not the same as `var` and is more like `IEnumerable<T>`. And for that...~ Per further discussion, it's a special use case entirely, and can be lowered to something even faster than an array with `IEnumerable<>`.
- For `IEnumerable<T>`, arrays should be used, IMO. The only promise is enumerability, and arrays satisfy that, and are cheaper than `List<T>`.
- Arrays already have a somewhat easier syntax - `SomeType[] x = [a, b, c];` - which is shorter than `List<SomeType> x = [a, b, c];` so I think it would be counter-productive to make `var` also do arrays, while leave `List<>` out in the cold.

Basically, I think `List<T>` is the best candidate, having considered all of the above.

**Reply by 333fred** [reply]
Most of those points don't strike home for me. For 1, I just don't think that's true. It feels like speculation through the lens of a personal bias. For 3, I don't think we should promise _any_ specific type. You asked for an `IEnumerable`, you get an `IEnumerable`. For 4, that's just irrelevant. You can already do `SomeType[] x = { a, b, c };` today, does that mean we should not have collection literals support creating arrays?

**Reply by TahirAhmadov** [reply]
1. I'm more than willing to hear the other side of that argument. What are those use cases? That conversation already happened and nobody suggested anything meaningful.
2. ---
3. Agreed on the promise; my point was that, in most cases, arrays should end up being used at the end.
4. It doesn't mean that, because a) collection literals are a unified syntax for all types, b) they support other features like spreads. I mean that collection literals have a good reason to support arrays regardless of the natural type discussion.

**Reply by CyrusNajmabadi** [reply]
Note: @jnm2 and i are working on a new formalization here.  So it's quite possible that we may go in a very different direction.

**Reply by CyrusNajmabadi** [reply]
> For IEnumerable<T>, arrays should be used, IMO. The only promise is enumerability, and arrays satisfy that, and are cheaper than List<T>.

Note: i'm more of a proponent of not promising *anything*.  You asked for IEnumerable, that's all you can necessarily be sure to get.

**Reply by TahirAhmadov** [reply]
> Note: i'm more of a proponent of not promising anything. You asked for IEnumerable, that's all you can necessarily be sure to get.

Not to repeat myself - in case you missed it, I wrote above: 
Agreed on the promise; my point was that, in most cases, arrays should end up being used at the end.

**Reply by CyrusNajmabadi** [reply]
>  arrays should end up being used at the end.

I personally do not like that as it would be easy to have people start taking a dependency on that.  I legit would prefer this be a type you cannot necessarily know about *at all* :)

**Reply by TahirAhmadov** [reply]
I can imagine some cases where an array is not what's used; for example, `foreach(int x in [1, 2, 3]) { ... }` can probably be optimized to an inline array, or even something without `IEnumerable<T>` altogether, or something like that. I meant, for most other use cases - passing it around, etc. - arrays are the best.

**Reply by CyrusNajmabadi** [reply]
>  I mean that collection literals have a good reason to support arrays regardless of the natural type discussion.

I'm not exactly certain i am understanding thsi point.  If it helps, collection literals *100%* support arrays, and consider them a first-class collection-like type that these literals want to work well with :)   So, if you're asking for that, then you're def going to get taht.  If you're asking for anything else, please lmk as i think i am missing something :)

**Reply by TahirAhmadov** [reply]
> I personally do not like that as it would be easy to have people start taking a dependency on that. I legit would prefer this be a type you cannot necessarily know about at all :)

Sure! If such a type can exist which does not incur any more costs compared to an array, I would be all for it.

**Reply by CyrusNajmabadi** [reply]
> Sure! If such a type can exist which does not incur any more costs compared to an array, I would be all for it.

Not only might it not incur more costs, it could incur *less* costs.  :)   For example, dedicated types for representing a "singleton enumerable" etc.

**Reply by TahirAhmadov** [reply]
> I'm not exactly certain i am understanding thsi point. If it helps, collection literals _100%_ support arrays, and consider them a first-class collection-like type that these literals want to work well with :) So, if you're asking for that, then you're def going to get taht. If you're asking for anything else, please lmk as i think i am missing something :)

I was responding to a point @333fred made, quoting:
"You can already do SomeType[] x = { a, b, c }; today, does that mean we should not have collection literals support creating arrays?"
My point was that, given that `SomeType[] x = [a, b, c];` is a little shorter than the `List<>` version, it makes more sense to make `var` be `List<>`. 
Obviously there are other benefits, such as being able to add/remove elements.

**Reply by TahirAhmadov** [reply]
> Not only might it not incur more costs, it could incur less costs. :) For example, dedicated types for representing a "singleton enumerable" etc.

Definitely, which is what I wrote about [here](https://github.com/dotnet/csharplang/discussions/7168#discussioncomment-5810804).

**Reply by CyrusNajmabadi** [reply]
for `foreach` we're simply going to reserve that we are allowed to do anything (including not even having a collection at all if we determine it's not necessary). :)

**Reply by jnm2** [reply]
> I personally do not like that as it would be easy to have people start taking a dependency on that. I legit would prefer this be a type you cannot necessarily know about at all :)

@CyrusNajmabadi Side note, a way to solve this without bloat could be something special in debug and just an array when optimizations are enabled.

**Reply by TahirAhmadov** [reply]
@jnm2 
> Side note, a way to solve this without bloat could be something special in debug and just an array when optimizations are enabled.

`ImmutableArray` ?

**Reply by jnm2** [reply]
> You can already do `SomeType[] x = { a, b, c };` today, does that mean we should not have collection literals support creating arrays?

@333fred While in agreement with where you land, I feel like `..` is a worthy feature :D

**Reply by jnm2** [reply]
@TahirAhmadov Potentially something unspeakable, if Cyrus really wants to drive home the "don't downcast this to a concrete type" lesson.

**Reply by TahirAhmadov** [reply]
@jnm2 
> @333fred While in agreement with where you land, I feel like `..` is a worthy feature :D

Definitely, also just `[]` are better than `{}` for collections, even if it's the same number of characters.

**Reply by 333fred** [reply]
> My point was that, given that SomeType[] x = [a, b, c]; is a little shorter than the List<> version, it makes more sense to make var be List<>.

I was taking this philosophy to absurdity to illustrate why I don't think the argument is persuasive. That array types are slightly shorter to type is an extremely weak argument as to why the natural type of collection literals should be List. The natural type should be the type that best expresses the semantics we decide to use, if we indeed _have_ a natural type at all. Whether one type expression has shorter syntax than another does not enter into it.

**Reply by TahirAhmadov** [reply]
@333fred the shortness is just one reason and I admit, a minor one. However, together with the other benefits of `List<T>`, there is a compelling case to make it the natural type. I still haven't heard any likely usage scenarios where `var` locals need to be immutable. And if they are arrays, they can be modified, just not grown, which seems to be a huge bummer; a solution which would leave almost everybody dissatisfied.
I guess, the reason I brought up the brevity of `SomeType[]` as compared to other collection types is to address the desires of the group of users who want `var` to be arrays. The only reason I can think of why they would want it to be like that (correct me if I'm wrong) is the brevity of writing `var x = [a, b, c];` and getting an array; in that context, the counter-argument that I made, that `SomeType[] x = [a, b, c];` is already slight shorter than other collection types, makes sense. I responded with a weak counter-point to the original weak argument, if you will.
PS. For what it's worth, there is an extremely popular language out there which has the `[a, b, c]` syntax with the natural type of a mutable/growable list. :) I'm half joking, but it would be helpful to provide this power to C# users, as well.

**Reply by 333fred** [reply]
> correct me if I'm wrong

Indeed, that is wrong. They expect it to be like that because arrays are already deeply understood by the language and they do not think that adding after initialization is a major use case. I'm not sure whether I agree or not; I trust the working group to think about it more deeply and come back to the LDM with an answer.

> I still haven't heard any likely usage scenarios where var locals need to be immutable

This depends entirely on how you use collections. For myself and every roslyn developer, we'd love it if they were, it would save us a ton of manually specifying `ImmutableArray`.

**Reply by alrz** [reply]
> Why else would you assign a collection to a local, other than to modify it?

In which case you probably want `var list = [];` which is illegal even with a natural type. I rarely seen an initialized list with further Add methods except when it smells in some ways or another. At least with the current syntax it's clear that it may be mutated later on.

I would rather have the diamond syntax to make it less verbose when you need to spell out the target-type.

Regardless of the natural type, I think overload resolution should be made always non-ambiguous, that is, having a preferred collection type as a tie-breaker, especially if there's no natural type.

(I think params already does this within the domain of supported types)

**Reply by CyrusNajmabadi** [reply]
> In which case you probably want var list = []; which is illegal 

We're working on designs where this would now be illegal.

**Reply by alrz** [reply]
So assuming list as the natural type, would this be ambiguous,
```cs
void M(T[])
void M(Span<T>)
```

while this is not?
```cs
void M(T[])
void M(List<T>)
```

**Reply by TahirAhmadov** [reply]
What about generic inference like `List<_> x = ["a", "b"];`

**Reply by CyrusNajmabadi** [reply]
> So assuming list as the natural type, would this be ambiguous,

No.  Those don't involve a natural type.  Those are target types.  And we would certainly both want this to work, and pick the "better" overload.  Note that we already established this in the `params span` working group.  Since `params Span<T>` is better than `params T[]`, the same would certainly hold for collection literals.

> ```
> void M(T[])
> void M(List<T>)
> ```

Also not a natural type situation.  :)

IMO, we should have a betterness path that has this order:

1. spans.  least overhead, can often be on the stack if small enough.
2. arrays.  minimal overhead, but on heap.
4. constructible type (open question if Collection-Initialize or Construct-form have any priority between themselves).  These produce concrete types, thus producing something which can have lower overhead to use (ops are normally non virtual, and iterating is often no overhead).
5. interface types.  These tend to be the least desirable.  All ops on them are virtual.  Iterating incurs allocations.

**Reply by CyrusNajmabadi** [reply]
> What about generic inference like List<_> x = ["a", "b"];

That would be a different feature (inferring generic args).  If we ever do that feature, i would expect it to work with collection literals.

**Reply by alrz** [reply]
>we should have a betterness path that has this order

I missed the change at https://github.com/dotnet/csharplang/pull/7171, I guess that's exactly what you described?
